### PR TITLE
Added smoother interior shadows

### PR DESCRIPTION
**Pull request by jarlarntzen**
Hi. I've added larger bitmaps treated with Gaussian blur to smooth out the shadow edges. In addition, there were some issues with too sharp transitions between the images.  On higher resolutions, this really makes a clear difference when flying in the cockpit.

### CODE REVIEW

**Comment by wlbragg** [comment]
Hum, I need to pull this and see it in action, it looks nice and the cube I created was by no means perfect in perspective or anything else for all that matters. Also I  would like the opinion of some of the other members of the project team including about the larger size.

**Comment by jarlarntzen** [comment]
@wlbragg @wkitty42 
I first thought of blurring the original shadows but this didn't turn out nice at all due to the low original resolution so I resolved to recreate them using the interior-shadow-cubemap-creator.blend camera rig instead, with the camera centered inside the cockpit. 
This also solves the issue of the original strange tail shadow where the tail would be strangely skewed/shifted to one side whilst flying directly away from the sun. 
Finally, the shadow maps are blurred more for objects further away (like the tail) and less for objects closer to the observer (like left, right and front wind screen).
selection_028

File size is 145kb, up from 41 kb of the original. For comparison, this is still far less than the grayscale_*.png (311 kB), default (1MB) and blue (1MB) series of interior images but still boasts a much more visible result. 

I'll try if flightgear likes grayscale PNGs for the next pull request.

**Comment by jarlarntzen** [comment]
@wlbragg @wkitty42 
Notice how much better the images are aligned in the new cube as compared to the old shadow cube-map. Additionally, the left hand window (left image) has a 2px line running down the right hand edge which is very clearly visible as it runs across the panel in low sun conditions.
shadow-orig
The same line is also present in the grayscale/default/blue image sets and is what I suspect is the cause of the dark vertical line suddenly tearing across the instrument glass reflections. I'll investigate this further and create another set of reflection images using the Blender camera rig.

**Comment by gilbertohasnofb** [comment]
I just tested these and I think these shadows look great, I love the blurred effect! As for the size, 145 kB total isn't bad at all, and I just tried optimizing them with optipng and I could get them down to 90.9 kB without any loss.

But I would say you have to fix `black_nx.png` because it's showing a bit of the propeller on the right-hand side of the panel, and this will be wrong all the time when the engine is on (which is the main state of the engine in a simulator :wink: ) as well as most of the time when the engine is off (since the propeller won't coincide with the shadow unless you get a lot of luck). So please just mirror the left side.

**Comment by jarlarntzen** [comment]
@gilbertohasnofb Great! `black_xn.png` is fixed. Don't know how to attach to an existing commit/pull request so I'm just adding it here..

Ill test out `optipng` next time!

black_nx

**Comment by jarlarntzen** [comment]
Found out how to add more commits now too :smile:

**Comment by wlbragg** [comment]
Looks great, thank you. So are we in agreement to add this now or are we going to wait for the reflection cubes?

**Comment by jarlarntzen** [comment]
I 'm struggling a bit with the reflection cubes now due to issues with Blender so you might not want to wait. I'll make them and test them soon enough,  though.

**Comment by gilbertohasnofb** [comment]
@jarlarntzen we plan to merge our repository into FGDATA on Tuesday, so it would be great to have this before that if you possibly can.

**Comment by jarlarntzen** [comment]
Hi.
@gilbertohasnofb  @wlbragg @wkitty42 : I've now added new reflections ( `default_*.png` ) and shadows on windows now (`grayscale_*.png`) which, incidentally, are identical to the file-set responsible for shadows elsewhere in the cockpit (`black_*.png`). The latter file sets can likely really just be combined into one set to save space - there is no real reason to have two identical sets.

Tested while flying in the morning, noon, afternoon, dusk, evening and night. There are still some sharp edges on the front wind-screen but that's due to geometry. I'll have a look at that later.

**Reflection cube-map**

default_reflection

**Comment by gilbertohasnofb** [comment]
Also, I would say that we should use the default 4X-CHF livery for the reflection map rather than this one.

**Comment by wlbragg** [comment]
> Although in this case grayscale_*.png is responsible for the shadows on the windows

There should be no active shadow on the windows only reflection. We removed the shadow effect from the windows because it was interfering with the visual of seeing the external rain falling. But you may be right that at one time it gs was used for that.

**Comment by jarlarntzen** [comment]
This is what I've done so far.

1. **Reflection**
The reflection cube-map is solved in grayscale. I used Gimp's Selective Gaussian Blur to smooth out noise from the rendering while keeping the details which helped reduce the total file size down to a nifty 415 kB. It it looks very convincing as a reflection in a window glass, even for liveries with colours. I will do further testing with more liveries but it may turn out to be the only reflection cube-map that's actually required. .eff-files modified for new shadow cube location

2. **Shadow**
Looks like we can easily do with just one black cube-map for both shadows on windows and on the rest of the interior. .eff files to be updated to allow for this.

3. **Instrument reflection**
I'm doing this one later as I havn't yet started to figure out how the instruments are hooked up to the effects.

As for testing this I'm a little confused. I'm not really sure how to run fgfs with just this aircraft as fgfs seems to do stuff like `Reading systems from /usr/share/games/flightgear/Aircraft/c172p/Systems/systems.xml` 
This is what I'm using to start the sim.
` /usr/bin/fgfs --aircraft-dir=/home/jarl/flightgear-git/c172p-detailed --airport=ENCN --language=en`

**Comment by wlbragg** [comment]
>It it looks very convincing as a reflection in a window glass, even for liveries with colours.

Good, that is what I had in mind when it became apparent that the different liveries needed to be accounted for (that we would have to use a b/w or grayscale for the reflection).

>one black cube-map for both shadows on windows and on the rest of the interior

I don't think your going to be able to do a shadow on the windows, the effects were never designed with this in mind although I did indeed use them that way. When I did it interfered with seeing raindrops,not the ones on the glass, but the rain falling outside the window. If you can figure out a away to do both, I would be all for it, I really like the idea of the glass showing a dynamic shadow. If you search the closed issues you will find a thread on the glass shadow and some pictures of it compared to the rain falling outside and when the choice was made to remove the glass shadow in favor of seeing rain again. If your so inspired, you can go back in history using GIT and check out the glass shadow code before we removed it and use it for testing.

>Instrument reflection

If your talking reflection and not shadow, we even had that discussion, I don't recall the outcome of that but there is also a closed issue with that discussion. Perhaps others involved with this project can shed some light on it. Shadows on instruments caused more issues than than gain but it wouldn't hurt to revisit it with a fresh mind.

**Comment by onox** [comment]
@wlbragg Are there any issues remaining that @jarlarntzen needs to fix before we can merge this PR?

**Comment by jarlarntzen** [comment]
@onox @wlbragg I still have left to make actual commits and push the new reflection. I'll also check out if the window shadow indeed interferes with rain, although I think not as the shadows did already show on the windscreen, I only swapped the referred cubemap (default_ png)  for the shadow one.

**Comment by onox** [comment]
Okay, I'll leave the reviewing to @wlbragg, he knows much more about the shadows than I do.

**Comment by wlbragg** [comment]
>Okay, I'll leave the reviewing to @wlbragg

I'll see where this is at late tonight and again tomorrow throughout the day.

>I'll also check out if the window shadow indeed interferes with rain,

We may be talking about two different things. Are you seeing a dynamic shadows that moves with the available sun angle and the glass? Or are you talking about static reflections on the glass?

**Comment by jarlarntzen** [comment]
Hi.
I've done some more testing but I've focused on the shadows. I've had difficulties running tests with the c172p-detailed from this repo so I've just run the test with the c172p at git://git.code.sf.net/p/flightgear/fgdata . I wasn't sure how to combine the c172-detailed from this repo with the FGdata repo from SourceForge so I have done the testing in the SF repo and simply copied over images to this repo before committing. This is something I'll have to learn for the next time.
In a later pull request I'll go through and rename all shadow-related files to `shadow_*.png` and `reflection_*.png` to reduce the confusion.

1. **Precipitation** 
Does not work outside any aircraft: UFO, c172p, f-14b. It does show on the glass though. Not having rain outside the UFO leads me to believe that this does not have anything to do with the glass shadow.

2. **Glass Shadow**
The glass shadow is dynamic (as it is in the rest of the interior) and works as it should. I haven't really done anything here beyond replacing the grayscale pngs with copies of the black pngs and adjusting the  `<opacity-cube-center type="vec3d" n="0">0.45732 0.0 0.1825</opacity-cube-center>` to the new values from Blender. 
fg-shadow on glass
3. **Reflections**
The reflections seem very plausible against all liveries. Seems going almost grayscale here works beautifully. In other good news, the reflection cubemap is now only 165 kB as opposed to ~1MB earlier. Again, I've only adjusted the opacity-cube-center values here. There seems to be no reflections on the front glass, although those would only show the interior floor.
fg-reflection with dark livery

**Comment by onox** [comment]
Just use `--fg-aircraft=`. FG should then prefer that path over the one in fgdata.

**Comment by jarlarntzen** [comment]
@onox: Thanks! Alas, it didn't work so well. FG didn't prefer 
neither 
`--fg-aircraft=` 
nor 
   `--aircraft-dir=path          Specify the exact directory to use for the
                                aircraft (normally not required, but may be
                                useful). Interpreted relatively to the current
                                directory. Causes the <path-cache> from
                                autosave_X_Y.xml, as well as --fg-aircraft and
                                the FG_AIRCRAFT environment variable to be
                                bypassed.`

fgfs keeps loading a mix of the c172p in the specified path along with the system c172p:
`Reading systems from /usr/share/games/flightgear/Aircraft/c172p/Systems/systems.xml
Initializing HUD Instrument
Reading sound sound from /home/jarl/flightgear-git/c172p-detailed/c172-sound.xml`


Not sure how to get around this, I'm afraid.. link the directory into the fgdata?
The result is that I just get to see the

**Comment by wlbragg** [comment]
Now that explains it, @jarlarntzen your working on an effect that is no longer a part of the project. The shadow system your using is only in the "current" c172p in fgdata. The one in the project "master", that will replace the one in fgdata, has no shadow system because it interferes with the visual of external rain falling. I don't want you to waist your time on something that currently doesn't exist. If you can make both "visual rain" and "glass shadows" work together, then of course we would put it back in. Just beware all the code your working on is outdated and is going to be difficult to merge into master because your working on the wrong branch. You need to be working on a copy of the master branch at this website, not the one in fgdata. It will require cherry picking to get only the needed parts back in to master. Everything else in the files you have has diverged so much from master that it would be close to impossible to merge it back in.

**Comment by wlbragg** [comment]
@jarlarntzen, do me a favor and start FG and see if from the internal view you can see rain falling outside the aircraft?

**Comment by wlbragg** [comment]
>your working on an effect that is no longer a part of the project

More accurately, your seeing effects that are no longer a part of the aircraft. The shadow effect is being pulled from fgdata and no longer exists in our project. I don't know what files your working on, did you do a clone of the master on this website?

**Comment by wlbragg** [comment]
I tink @onox meant
--aircraft-dir="fill this in with your path to your c172p"

**Comment by wlbragg** [comment]
> I wasn't sure how to combine the c172-detailed from this repo with the FGdata repo from SourceForge so I have done the testing in the SF repo and simply copied over images to this repo before committing

Sorry I missed this part, but what your doing may be a problem because the systems are so divergent from one to the other that your lucky it works. Because your dealing with images and a system that hasn't changed too much your getting away with it, but as I have explained, not entirely, because you are showing effects that no longer exist.

How do you launch FG and what is you OS?

**Comment by jarlarntzen** [comment]
Quoting myself:

> **Precipitation**
Does not work outside any aircraft: UFO, c172p, f-14b. It does show on the glass though. Not having rain outside the UFO leads me to believe that this does not have anything to do with the glass shadow.

This is my FG version:

```
/usr/bin/fgfs --version
platform default fg_root = /usr/share/games/flightgear
Reading global preferences
Finished Reading global preferences
Reading user settings from /home/jarl/.fgfs/autosave_2016_1.xml
creating catalog from:Path "/home/jarl/.fgfs/Aircraft/org.flightgear.official"
No user specified aircraft, using default
Loading aircraft -set file from:/usr/share/games/flightgear/Aircraft/c172p/c172p-set.xml
Found language resource for: en_US
Reading localized strings for 'en' from /usr/share/games/flightgear/Translations/en/sys.xml
Reading localized strings for 'en' from /usr/share/games/flightgear/Translations/en/atc.xml
FlightGear version: 2016.1.0
Revision: none
Build-Id: none
FG_ROOT=/usr/share/games/flightgear
FG_HOME=/home/jarl/.fgfs
FG_SCENERY=
SimGear version: 2016.1.0
OSG version: 3.2.0
PLIB version: 185
Saving user settings to /home/jarl/.fgfs/autosave_2016_1.x
```

- I'm on Ubuntu 14.04 LTS.
- FGData is pulled daily from sourceforge
- fgfs binary is from the ppa at launchpad.


I launch via command line to easily view all error messages.

I do think simply using the new PNGs will work well but since my system evidently isn't set up properly, it would be great if you @wlbragg could do a pull and test this or perhaps tell me how my system should Ideally be set up for testing the **Juanvvc/c172p-detailed** repo. I am in no rush, better to do it right the first time.

**Comment by wlbragg** [comment]
>Does not work outside any aircraft: UFO, c172p, f-14b. It does show on the glass though.

rain
Yes it does work!

It is possible that FG next is messed up or that the ufo is a special circumstance and that the f-14b had the same problem we did and never noticed it. I am using a FG nightly from prior to the particle effects being broken.

I also vividly remember the conversation and was the person that did all the testing on the issue. So I can assure you it was because of the effects and the way the glass was arranged.

I need to point you to the closed issue regarding this, as the reason I believe was included in that discussion. I just can't off the top of my head remember if it was the two sided glass issue or the normal's or just the duplicate effects.

I'll see if I can find it and link you to it.

**Comment by wlbragg** [comment]
As far as you system setup,

So if you pulled from this website GIT and assuming you can get the following to work, you should be good to go.

>I think @onox meant
--aircraft-dir="fill this in with your path to your GIT master c172p"

**Comment by wlbragg** [comment]
>No user specified aircraft, using default
Loading aircraft -set file from:/usr/share/games/flightgear/Aircraft/c172p/c172p-set.xml

You have to get around this.

**Comment by jarlarntzen** [comment]
@wlbragg : There's a fair chance my FGdata is messed up but I probably will not have time to figure out this before end of this week due to work etc.
I did pull from this website: https://github.com/Juanvvc/c172p-detailed/ so, at least what I have pushed has originated the right place.

I'll try to get --aircraft-dir to work properly but seems like I have to learn why it still prefers the installed files over the --aircraft-dir folder.

When were the particle effects broken? Is this due to the fgfs binary or fgdata?

Would be great if you want to have a go and test the PNG-files, I think it should be fairly straightforward to simply download them all into the correct folder, and it should work.

**Comment by wlbragg** [comment]
>Is this due to the fgfs binary or fgdata?

fgfs binary

>Would be great if you want to have a go and test the PNG-files

We should be able to get that in, thanks.

**Comment by onox** [comment]
In `.fgfsrc` I have:

```
--fg-aircraft=path/to/folder/containing/symlinks/which/are/just/the/aircraft/names
--fg-aircraft=path/to/c172p
--fg-aircraft=path/to/c182
```

etc.

**Comment by onox** [comment]
@jarlarntzen Just nuke that old c172p thoroughly temporarily by renaming it or something. Try to start fg and make sure it cannot find the c172p. Then get the git clone of our development c172p.

**Comment by onox** [comment]
You can also add `--fg-root=path/to/your/custom/fgdata` to avoid Ubuntu's fgdata.

**Comment by jarlarntzen** [comment]
> I need to point you to the closed issue regarding this, as the reason I believe was included in that discussion. I just can't off the top of my head remember if it was the two sided glass issue or the normal's or just the duplicate effects.

Thanks. If you happen to come across the issue I'm interested.

**Comment by wlbragg** [comment]
I couldn't find it, but it is easy enough to test, when your ready I can give you some pointers on where to start at least.

**Comment by jarlarntzen** [comment]
@wlbragg @onox Thanks for your help so far. I'll try out your suggestions at a later stage.

For now, I am not realistically capable of testing this in the right environment for at least a week or two and I doubt I will be able to provide a fully tested pull request until late next week at the earliest.

So, at the moment, all I'm offering are the bitmaps for you to download and test and possibly decide to commit if time is of the essence. I think, at least, they are of a pretty decent quality and will provide an improvement for the c172p detailed.

Thanks.

**Comment by wlbragg** [comment]
OK, on this I don't want to add anything that is not correct and applicable now. I am a bit punch drunk after all the blend work I did today and need to take a fresh look at this in the morning if there is time before the merge.

Specifically I think all that is here that is ready to merge is the interior shadow cubes which is the 
black_XX.png's
commit - Added smoother interior shadows 	bf3a357
and the propeller fix
commit - Removed visible propeller tip on black_nx.png 	03af917
and the last commit
commit - 	Updated reflections with grayscale versions. … 	6da691d
simply replaces the current red reflection cube with greyscale?
About this, I really like it, it is what we need because it is livery independent, but I also want to keep a red and blue version if possible, even if we only keep them in dev/.

Bottom line is I think were OK to merge this.

**Comment by wlbragg** [comment]
OK, I am not sure how or where this happened but the grayscale_XX.png's are all black and white, I would like them to be pure greyscale without any editing, if you know what I mean (as opposed to the default_xx.png's you made)..
Here is a list of what I would like to see in this commit if you have the time and don't mind.
1) Replace these current greyscale (they currently are blurred edged b/w) with a true greyscale with your new proportions.
2) Create a red livery set with your new proportions and name them red_xx.png.
3) Replace the existing blue livery set with a new blue set with your new proportions.
Leave the new black and the new default sets alone.

I would really appreciate this if you don't mind doing it. Your work is far superior to what we have.

**Comment by onox** [comment]
> (I also have to gain a better understanding of git through a tutorial or similar.)

Try https://www.atlassian.com/git/

**Comment by onox** [comment]
Or https://try.github.io/

**Comment by onox** [comment]
@wlbragg Don't hesitate to postpone this until after I have branched `release/2016.1` if it requires a bit more work to get it done right.

**Comment by wlbragg** [comment]
When are you branching release/2016.1?

**Comment by onox** [comment]
Around 1800 UTC+1, that's in about 10 hours.

**Comment by wlbragg** [comment]
I was getting ready to use part of this but I still have a question.

Why on the default (greyscale) cube did you not include any gauges in the instrument panel, only holes? I'm not totally convinced that is what we want for a reflection.

I'm interested in everyone's opinion on this.

As of this moment the only cubes I am certain about including are the black_xx.png's.

**Comment by wlbragg** [comment]
I'll try to check back in prior to 1800 UTC+1 and see if there is any changes or responses.

**Comment by jarlarntzen** [comment]
@onox Thanks for the pointers. Very educational. 

@wlbragg 
> 1) Replace these current greyscale (they currently are blurred edged b/w) with a true greyscale with your new proportions.

My thought here were to only change the actual bitmaps in order to create the correct effects without impacting any code, as such, and to make merging involve less files and no deletions. (I.e. by keeping the same file names but only changing the content just in case some code refers directly to the actual file names)   Later my plan is to go modify the file names . The files are called `greyscale_*.png` but should really be called `shadow_glass-*.png` because that's what they are actually used for. 
Also, ideally, there would be no need for a separate `shadow_glass-*.png` since all shadows can be handled by a single set of `shadow_interior-*.png` files.

The same goes for `default_*.png` which should be renamed to `reflection_default*.png`, again because that's what they are really used for.

> 2) Create a red livery set with your new proportions and name them red_xx.png.
No problem. Only `reflection_red_livery-*.png` would be more correct.


> 3) Replace the existing blue livery set with a new blue set with your new proportions.
Leave the new black and the new default sets alone.
Again, no problem. 

>Why on the default (greyscale) cube did you not include any gauges in the instrument panel, only holes? I'm not totally convinced that is what we want for a reflection.

This is simply what Blender gives me and I'm afraid I do not know how to display the instruments now. Even so, as it turns out, due to the angle of the front windscreen, I've yet to see reflections of anything but the floor in it. The instrument panel is only barely visible in the side glass and only really when the glass is fogged over.

I'll have to postpone any further work on this now until after the release branch and will return with a fresh pull request (or a full commit to this one) which will include all necessary changes: 
All requested files added. (shadow, reflection_default (grayscale), reflection_blue, reflection_red)
File renaming.
File deletions.
File edits.
Fully tested.
